### PR TITLE
[HTTP Endpoint] Add support for combined interfaces

### PR DESCRIPTION
**Pull request by mjacoby**
* support for contextualized HTTP request mappings
* HTTP endpoint now serving SubmodelInterface requests conxtualized with aas (e.g. /shells/{aasIdentifier}/aas/submodels/{submodelIdentifier}/submodel)

### CODE REVIEW

File: model/src/test/java/de/fraunhofer/iosb/ilt/faaast/service/model/AASFull.java [hunk 1]
```diff
@@ -136,6 +136,13 @@ public static AssetAdministrationShell createAAS1() {
                                 .idType(KeyType.IRI)
                                 .build())
                         .build())
+                .submodel(new DefaultReference.Builder()
```
Comment (mjacoby) [inline model/src/test/java/de/fraunhofer/iosb/ilt/faaast/service/model/AASFull.java:139]: .json file was not updated - no failing tests???

File: endpoint/http/src/main/java/de/fraunhofer/iosb/ilt/faaast/service/endpoint/http/request/mapper/DeleteAllAssetLinksByIdRequestMapper.java [hunk 2]
```diff
@@ -21,32 +21,26 @@
...
 
 
 /**
  * class to map HTTP-DELETE-Request path: lookup/shells/{aasIdentifier}
  */
 public class DeleteAllAssetLinksByIdRequestMapper extends RequestMapper {
 
-    private static final HttpMethod HTTP_METHOD = HttpMethod.DELETE;
-    private static final String PATTERN = "^lookup/shells/(.*)$";
+    private static final String AAS_ID = "aasId";
```
Comment (JensMueller2709) [inline endpoint/http/src/main/java/de/fraunhofer/iosb/ilt/faaast/service/endpoint/http/request/mapper/DeleteAllAssetLinksByIdRequestMapper.java:32]: Might be useful to create a static helper class or move to super class for such constants. Same for `submodelId`, `submodelelementId`, `conceptId` and `packageId`

File: endpoint/http/src/main/java/de/fraunhofer/iosb/ilt/faaast/service/endpoint/http/request/mapper/RequestMapper.java [hunk 3]
```diff
@@ -30,10 +37,23 @@ public abstract class RequestMapper {
 
     protected ServiceContext serviceContext;
```
Comment (JensMueller2709) [inline endpoint/http/src/main/java/de/fraunhofer/iosb/ilt/faaast/service/endpoint/http/request/mapper/RequestMapper.java:38]: Could be set also to `final`

File: endpoint/http/src/main/java/de/fraunhofer/iosb/ilt/faaast/service/endpoint/http/request/mapper/RequestMapper.java [hunk 4]
```diff
@@ -30,10 +37,23 @@ public abstract class RequestMapper {
...
     protected HttpJsonDeserializer deserializer;
```
Comment (JensMueller2709) [inline endpoint/http/src/main/java/de/fraunhofer/iosb/ilt/faaast/service/endpoint/http/request/mapper/RequestMapper.java:39]: Could be set also to `final`

File: endpoint/http/src/main/java/de/fraunhofer/iosb/ilt/faaast/service/endpoint/http/request/mapper/SetSubmodelElementValueByPathRequestMapper.java [hunk 5]
```diff
@@ -33,31 +35,41 @@
...
+                && Objects.equals(httpRequest.getQueryParameter(QUERY_PARAMETER_CONTENT), QUERY_PARAMETER_CONTENT_VALUE);
+    }
+
+
+    @Override
+    public SetSubmodelElementValueByPathRequest doParse(HttpRequest httpRequest, Map<String, String> urlParameters, OutputModifier outputModifier) throws InvalidRequestException {
+        final List<Key> path = ElementPathHelper.toKeys(EncodingHelper.urlDecode(urlParameters.get(SUBMODEL_ELEMENT_PATH)));
+        final Identifier identifier = IdentifierHelper.parseIdentifier(EncodingHelper.base64Decode(urlParameters.get(SUBMODEL_ID)));
         return SetSubmodelElementValueByPathRequest.builder()
-                .id(identifier)
```
Comment (JensMueller2709) [inline endpoint/http/src/main/java/de/fraunhofer/iosb/ilt/faaast/service/endpoint/http/request/mapper/SetSubmodelElementValueByPathRequestMapper.java:60]: Add `submodelId` here?

File: model/src/main/java/de/fraunhofer/iosb/ilt/faaast/service/model/request/OutputModifierConstraints.java [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+    private boolean supportsExtent;
+    private boolean supportsLevel;
+    private Set<Content> supportedContentModifiers;
+    public static final OutputModifierConstraints DEFAULT = new OutputModifierConstraints(true, false, Content.values());
+    public static final OutputModifierConstraints NONE = new OutputModifierConstraints(false, false);
+    public static final OutputModifierConstraints SUBMODEL = new OutputModifierConstraints(true, true, Content.NORMAL, Content.VALUE, Content.REFERENCE, Content.PATH);
+    public static final OutputModifierConstraints SUBMODEL_ELEMENT = new OutputModifierConstraints(true, true, Content.NORMAL, Content.VALUE, Content.REFERENCE, Content.PATH);
+    public static final OutputModifierConstraints ASSET_ADMINISTRATION_SHELL = new OutputModifierConstraints(false, false, Content.NORMAL, Content.REFERENCE);
+
+    public OutputModifierConstraints() {
```
Comment (JensMueller2709) [inline model/src/main/java/de/fraunhofer/iosb/ilt/faaast/service/model/request/OutputModifierConstraints.java:41]: May add javadoc with hints for `this.supportsExtent = true` and `this.supportsLevel = false`

**Review by JensMueller2709 — CHANGES_REQUESTED** [review]

File: core/src/main/java/de/fraunhofer/iosb/ilt/faaast/service/request/handler/SubmodelInterfaceRequestHandler.java [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+     *             not belong to AAS
+     */
+    protected void validateSubmodelWithinAAS(T request) throws ResourceNotFoundException {
+        if (request.getAasId() != null) {
+            Reference submodelRef = ReferenceHelper.toReference(request.getSubmodelId(), Submodel.class);
+            if (((AssetAdministrationShell) persistence.get(
+                    request.getAasId(),
+                    new OutputModifier.Builder()
+                            .level(Level.CORE)
+                            .build())).getSubmodels().stream().noneMatch(x -> Objects.equals(x, submodelRef))) {
```
[the remark this answers is not in the document]
Comment (mjacoby) [reply core/src/main/java/de/fraunhofer/iosb/ilt/faaast/service/request/handler/SubmodelInterfaceRequestHandler.java:72]: Replaced `Objects.equals(...)` with `ReferenceHelper.isEqualsIgnoringKeyType(...)`

File: core/src/main/java/de/fraunhofer/iosb/ilt/faaast/service/request/handler/SubmodelInterfaceRequestHandler.java [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+
+/**
+ * Abstract base class for handling request that are part of the Submodel
+ * Interface. If request is made in the context of an AAS, this handler
+ * validates that the AAS exists and the requested submodel is part of that AAS.
+ *
+ * @param <T> actual type of the request
+ * @param <U> actual type of the response
+ */
+public abstract class SubmodelInterfaceRequestHandler<T extends SubmodelInterfaceRequest<U>, U extends Response> extends RequestHandler<T, U> {
```
[the remark this answers is not in the document]
Comment (mjacoby) [reply core/src/main/java/de/fraunhofer/iosb/ilt/faaast/service/request/handler/SubmodelInterfaceRequestHandler.java:41]: Will keep `SubmodelInterface` as this referes to the set of API calls with exactly this name but added `Abstract...` before to highlight it is an abstract base class.

File: endpoint/http/src/main/java/de/fraunhofer/iosb/ilt/faaast/service/endpoint/http/request/mapper/SetSubmodelElementValueByPathRequestMapper.java (hunk 5 above)
Comment (mjacoby) [reply endpoint/http/src/main/java/de/fraunhofer/iosb/ilt/faaast/service/endpoint/http/request/mapper/SetSubmodelElementValueByPathRequestMapper.java:60]: This happens in parent class https://github.com/FraunhoferIOSB/FAAAST-Service/blob/139197116ce904bf9245fba390fa2328221ace0d/endpoint/http/src/main/java/de/fraunhofer/iosb/ilt/faaast/service/endpoint/http/request/mapper/SubmodelInterfaceRequestMapper.java#L135

**Review by JensMueller2709 — APPROVED** [review]

**Pull request merged**